src/app/AttributePathParams.h

```diff
@@ -65,10 +63,30 @@ struct AttributePathParams
     inline bool HasWildcardAttributeId() const { return mAttributeId == kInvalidAttributeId; }
     inline bool HasWildcardListIndex() const { return mListIndex == kInvalidListIndex; }
 
-    EndpointId mEndpointId   = kInvalidEndpointId;
-    ClusterId mClusterId     = kInvalidClusterId;
-    AttributeId mAttributeId = kInvalidAttributeId;
-    ListIndex mListIndex     = kInvalidListIndex;
+    bool IsAttributePathSupersetOf(const AttributePathParams & other) const
+    {
+        VerifyOrReturnError(HasWildcardEndpointId() || mEndpointId == other.mEndpointId, false);
+        VerifyOrReturnError(HasWildcardClusterId() || mClusterId == other.mClusterId, false);
+        VerifyOrReturnError(HasWildcardAttributeId() || mAttributeId == other.mAttributeId, false);
+        VerifyOrReturnError(HasWildcardListIndex() || mListIndex == other.mListIndex, false);
+
+        return true;
+    }
+
+    bool IsAttributePathSupersetOf(const ConcreteAttributePath & other) const
+    {
+        VerifyOrReturnError(HasWildcardEndpointId() || mEndpointId == other.mEndpointId, false);
+        VerifyOrReturnError(HasWildcardClusterId() || mClusterId == other.mClusterId, false);
+        VerifyOrReturnError(HasWildcardAttributeId() || mAttributeId == other.mAttributeId, false);
+
+        return true;
+    }
+
+    ClusterId mClusterId         = kInvalidClusterId;   // uint32
+    AttributeId mAttributeId     = kInvalidAttributeId; // uint32
+    AttributePathParams * mpNext = nullptr;             // pointer width (32/64 bits)
```

Comment:
So now some places we have linked lists of AttributePathParams and in other places (e.g. when setting up a ReadClient) we have arrays of them?  That's pretty confusing, and is likely to mislead someone into walking lists when they should be going through the array or vice versa.

It would be better, I think, to have an `AttributePath` class which either inherits from AttributePathParams or has it as a member and has the additional `mpNext`.  That would also have the benefit that on the server at least the terminology would be much more aligned with the spec..

Reply:
updated, thanks